File: BinaryTree/12.InorderOfATreeByMorrisTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        if(root == NULL) {
            return ans;
        }
        
        TreeNode * curr = root;
        while(curr != NULL) {
            if(curr->left == NULL) {
                ans.push_back(curr->val);
                curr = curr->right;
            }else{
                TreeNode * iop = curr->left;
                while(iop->right != NULL && iop->right != curr) {
                    iop = iop->right;
                }
                if(iop->right == NULL) {
                    iop->right = curr;
                    curr = curr -> left;
                }else{
                    iop->right = NULL;
                    ans.push_back(curr->val);
                    curr = curr->right;
                }
            }  
        }
        return ans;
    }
};

File: BinaryTree/12.InorderOfATreeByMorrisTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // Space - O(1) /// not even recursion stack
    // Time - O(4N) or O(N) As we are visiting node atmost 2 times only
    public List<Integer> inorderTraversal(TreeNode root) {
        
        List<Integer> retVal = new ArrayList<>();
        if(root == null) return retVal;
        
        TreeNode curr = root;
        while(curr != null) {
            if(curr.left == null){
                // Print
                retVal.add(curr.val);
                // go to right
                curr = curr.right;
            }else{ // when curr.left != null
                // see if left is processed or not(inorderPredecessor is left ka rightMost)
                TreeNode iop = curr.left;
                while(iop.right != null && iop.right != curr) {
                    iop = iop.right;
                }
                // if left is not processed
                if(iop.right == null) {
                    // inorderPredecessor.right = curr
                    iop.right = curr;
                    // and then go to left
                    curr = curr.left;
                }
                // if left is processed
                else{ // when iop.right = curr
                    // inorderPredecessor.right = null
                    iop.right = null;
                    // Print
                    retVal.add(curr.val);
                    // go to right
                    curr = curr.right;
                }
            }
        }
        return retVal;
    }
}